### ISSUE TITLE
metrics中auc的实现问题

### ISSUE THREAD

**Issue by zbt78** [DongHande/AutoDebias#6]
hello，
`def auc(vector_predict, vector_true, device = 'cuda'): 
    pos_indexes = torch.where(vector_true == 1)[0].to(device)
    pos_whe=(vector_true == 1).to(device)
    sort_indexes = torch.argsort(vector_predict).to(device)
    rank=torch.zeros((len(vector_predict))).to(device)
    rank[sort_indexes] = torch.FloatTensor(list(range(len(vector_predict)))).to(device)
    rank = rank * pos_whe
    auc = (torch.sum(rank) - len(pos_indexes) * (len(pos_indexes) - 1) / 2) / (len(pos_indexes) * (len(vector_predict) - len(pos_indexes)))
    return auc.item()`
vector_predict是所有用户对物品的预测，但是计算auc时应该要分用户进行计算，这样混着计算没什么用吧？

**Comment by zbt78** [comment]
希望能够得到您的答复

**Comment by DongHande** [comment]
AUC有两种计算方法，一种是对全量的（相当于把这个问题就看做是全量的排序问题），另一种是对每个用户计算然后求平均，两者皆可。我们这么计算AUC，主要是follow之前几篇论文的相关的实验设置。

**Comment by zbt78** [comment]
谢谢！
还有一个问题是：在把评分转化成隐形反馈时，用1和-1来替换，而不是用传统的1和0，请问有什么区别吗？

**Comment by DongHande** [comment]
没有什么区别。我们实验在优化模型的时候，用的是MSE loss，对于这个loss来说，1/-1 相比于 1/0 只是做了个放缩。评估指标方面：使用论文中的公式（27）计算NLL会有一些区别（看一下论文里面NLL的计算公式很容易就明白了，这里就不解释了），AUC和NDCG也没影响。

**Comment by zbt78** [comment]
好的，谢谢你
